agent: segfault in cecManager.getEndpoint()

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

equal or higher than v1.16.5 and lower than v1.17.0

### What happened?

We've seen agent segfaults on versions 1.16.2 and 1.16.5 with the following approximate stack trace:

```
/go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/job/job.go:159 +0x174
created by github.com/cilium/hive/job.(*group).Start.func1 in goroutine 1
/go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/job/timer.go:131 +0x448
github.com/cilium/hive/job.(*jobTimer).start(0xc00243b6e0, {0x452ea10, 0xc002d97380}, 0xc00354a7d0?, {0x4536c60, 0xc00243b440}, {{{0xc0008793a0, 0x1, 0x1}}, 0xc000e5b850, ...})
/go/src/github.com/cilium/cilium/pkg/ciliumenvoyconfig/cec_reconciler.go:306 +0x197
github.com/cilium/cilium/pkg/ciliumenvoyconfig.(*ciliumEnvoyConfigReconciler).syncHeadlessService(0xc002464000, {0xc011f67d78?, 0x49ef6f?})
/go/src/github.com/cilium/cilium/pkg/ciliumenvoyconfig/cec_manager.go:180 +0x212
github.com/cilium/cilium/pkg/ciliumenvoyconfig.(*cecManager).syncCiliumEnvoyConfigService(0xc0002969a0, {0xc00597a510, 0xe}, {0xc00a7ab8c0, 0x1b}, 0xc0254a9f80?)
/go/src/github.com/cilium/cilium/pkg/ciliumenvoyconfig/cec_manager.go:252 +0xd3
github.com/cilium/cilium/pkg/ciliumenvoyconfig.(*cecManager).getEndpoint(0xc00a7ab8c0?, {0xc00597a600, 0xe}, {0xc00a7ab900, 0x1b})
goroutine 1555 [running]:
[signal SIGSEGV: segmentation violation code=0x1 addr=0xa0 pc=0x2c3fc73]
panic: runtime error: invalid memory address or nil pointer dereference
```

When this occurs, CEC reconciliation stops on the node and ClusterIP services become unreachable (though direct-dialing pods still works).

### How can we reproduce the issue?

1. Install cilium 1.16.5 via helm
1. Create at least one CEC
1. Create a large number of endpoints and churn them constantly
1. Wait for the reconciliation loop to segfault

### Cilium Version

```
Client: 1.16.5 ad688277 2024-12-12T20:18:31+00:00 go version go1.22.10 linux/amd64
Daemon: 1.16.5 ad688277 2024-12-12T20:18:31+00:00 go version go1.22.10 linux/amd64
```

### Kernel Version

`Linux 6.1.119 #1 SMP PREEMPT_DYNAMIC Fri Dec 20 22:32:32 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux`
(via Bottlerocket)

### Kubernetes Version

1.29 EKS

### Regression

No

### Sysdump

_No response_

### Relevant log output

```shell

```

### Anything else?

_No response_

### Cilium Users Document

- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

--- a/pkg/ciliumenvoyconfig/cec_manager.go
+++ b/pkg/ciliumenvoyconfig/cec_manager.go
@@ -272,7 +272,7 @@ func (r *cecManager) getEndpoint(serviceName string, serviceNamespace string) (*
 	iter := store.IterKeys()
 	for iter.Next() {
 		e, _, _ := store.GetByKey(iter.Key())
-		if e.EndpointSliceID.ServiceID.Name == serviceName &&
+		if e != nil && e.EndpointSliceID.ServiceID.Name == serviceName &&
 			e.EndpointSliceID.ServiceID.Namespace == serviceNamespace {
 			if res == nil {
 				res = e.DeepCopy()

